File: src/services/aiService.js
import * as tf from '@tensorflow/tfjs';

let sentimentModel = null;
let isModelLoaded = false;

// Mock sentiment analysis using TensorFlow.js patterns
// In a real implementation, you'd load a pre-trained model
export const initializeAI = async () => {
  try {
    console.log('Initializing AI models...');
    
    // Simulate model loading time
    await new Promise(resolve => setTimeout(resolve, 2000));
    
    // Create a simple mock model for demonstration
    sentimentModel = {
      predict: (text) => {
        // Simple keyword-based sentiment analysis for demo
        const positiveWords = ['good', 'great', 'happy', 'love', 'awesome', 'amazing', 'excellent', 'wonderful', 'fantastic', 'perfect', 'best', 'better', 'beautiful', 'brilliant', 'cheerful', 'delighted', 'excited', 'grateful', 'hopeful', 'inspired', 'joyful', 'optimistic', 'pleased', 'satisfied', 'thrilled'];
        const negativeWords = ['bad', 'sad', 'hate', 'terrible', 'awful', 'horrible', 'worst', 'angry', 'depressed', 'anxious', 'stressed', 'worried', 'frustrated', 'disappointed', 'upset', 'miserable', 'devastated', 'heartbroken', 'lonely', 'overwhelmed', 'exhausted', 'hopeless', 'discouraged', 'bitter', 'resentful'];
        const stressWords = ['exam', 'test', 'deadline', 'assignment', 'project', 'study', 'finals', 'midterm', 'homework', 'workload', 'pressure', 'busy', 'overwhelmed', 'stressed', 'tired', 'exhausted', 'rush', 'panic', 'behind', 'late'];
        
        const words = text.toLowerCase().split(' ');
        let positiveScore = 0;
        let negativeScore = 0;
        let stressScore = 0;
        
        words.forEach(word => {
          if (positiveWords.includes(word)) positiveScore++;
          if (negativeWords.includes(word)) negativeScore++;
          if (stressWords.includes(word)) stressScore++;
        });
        
        // Determine primary sentiment
        if (stressScore > 0 && stressScore >= positiveScore) {
          return { label: 'Stressed', confidence: Math.min(0.7 + stressScore * 0.1, 0.95) };
        } else if (negativeScore > positiveScore) {
          return { label: 'Negative', confidence: Math.min(0.6 + negativeScore * 0.1, 0.95) };
        } else if (positiveScore > negativeScore) {
          return { label: 'Positive', confidence: Math.min(0.6 + positiveScore * 0.1, 0.95) };
        } else {
          return { label: 'Neutral', confidence: 0.5 };
        }
      }
    };
    
    isModelLoaded = true;
    console.log('AI models loaded successfully');
    
  } catch (error) {
    console.error('Failed to initialize AI models:', error);
    throw error;
  }
};

export const analyzeSentiment = async (text) => {
  if (!isModelLoaded || !sentimentModel) {
    throw new Error('AI models not initialized');
  }
  
  try {
    // Simulate processing time
    await new Promise(resolve => setTimeout(resolve, 100));
    
    const result = sentimentModel.predict(text);
    
    return {
      label: result.label,
      confidence: result.confidence,
      text: text,
      timestamp: new Date().toISOString()
    };
  } catch (error) {
    console.error('Sentiment analysis failed:', error);
    throw error;
  }
};

export const analyzeTypingPatterns = (keystrokes) => {
  // Analyze typing speed, pauses, and patterns
  const avgSpeed = keystrokes.length > 0 ? 
    keystrokes.reduce((sum, k) => sum + (k.speed || 0), 0) / keystrokes.length : 0;
  
  const longPauses = keystrokes.filter(k => k.pause > 2000).length;
  const rapidTyping = keystrokes.filter(k => k.speed > 5).length;
  
  let emotionalState = 'neutral';
  let confidence = 0.5;
  
  if (longPauses > 3) {
    emotionalState = 'contemplative';
    confidence = 0.7;
  } else if (rapidTyping > 5) {
    emotionalState = 'agitated';
    confidence = 0.8;
  } else if (avgSpeed < 2) {
    emotionalState = 'tired';
    confidence = 0.6;
  }
  
  return {
    emotionalState,
    confidence,
    metrics: {
      averageSpeed: avgSpeed,
      longPauses,
      rapidTyping,
      totalKeystrokes: keystrokes.length
    }
  };
};

export const generatePersonalizedIntervention = (mood, userPreferences = {}) => {
  const interventions = {
    negative: [
      {
        type: 'breathing',
        title: 'Calm Your Mind',
        description: 'Try the 4-7-8 breathing technique: Inhale for 4, hold for 7, exhale for 8',
        duration: '3 minutes',
        effectiveness: 0.85
      },
      {
        type: 'cognitive',
        title: 'Reframe Your Thoughts',
        description: 'Ask yourself: "Is this thought helping me right now? What would I tell a friend?"',
        duration: '5 minutes',
        effectiveness: 0.75
      },
      {
        type: 'physical',
        title: 'Take a Walk',
        description: 'A short walk can shift your perspective and boost your mood',
        duration: '10 minutes',
        effectiveness: 0.80
      }
    ],
    stressed: [
      {
        type: 'time-management',
        title: 'Priority Check',
        description: 'List your top 3 priorities for today. Focus on just one at a time.',
        duration: '5 minutes',
        effectiveness: 0.90
      },
      {
        type: 'relaxation',
        title: 'Progressive Muscle Relaxation',
        description: 'Tense and release each muscle group, starting from your toes',
        duration: '10 minutes',
        effectiveness: 0.85
      }
    ],
    anxious: [
      {
        type: 'grounding',
        title: '5-4-3-2-1 Technique',
        description: 'Name 5 things you see, 4 you hear, 3 you touch, 2 you smell, 1 you taste',
        duration: '3 minutes',
        effectiveness: 0.88
      }
    ]
  };
  
  const moodInterventions = interventions[mood.toLowerCase()] || interventions.negative;
  
  // Personalize based on user preferences
  if (userPreferences.preferredDuration) {
    return moodInterventions.filter(i => 
      parseInt(i.duration) <= userPreferences.preferredDuration
    );
  }
  
  return moodInterventions;
};

export const predictMoodTrend = (moodHistory) => {
  if (moodHistory.length < 7) {
    return { trend: 'insufficient_data', confidence: 0 };
  }
  
  // Simple trend analysis based on recent mood patterns
  const recentMoods = moodHistory.slice(-7);
  const moodScores = recentMoods.map(entry => {
    switch (entry.mood) {
      case 'positive': return 2;
      case 'neutral': return 1;
      case 'negative': return -1;
      case 'stressed': return -2;
      default: return 0;
    }
  });
  
  const averageScore = moodScores.reduce((sum, score) => sum + score, 0) / moodScores.length;
  const trend = averageScore > 0.5 ? 'improving' : averageScore < -0.5 ? 'declining' : 'stable';
  
  // Calculate trend strength
  const variance = moodScores.reduce((sum, score) => sum + Math.pow(score - averageScore, 2), 0) / moodScores.length;
  const confidence = Math.min(Math.abs(averageScore) + (1 / (1 + variance)), 0.95);
  
  return {
    trend,
    confidence,
    averageScore,
    recommendation: generateTrendRecommendation(trend, averageScore)
  };
};

const generateTrendRecommendation = (trend, score) => {
  if (trend === 'declining') {
    return {
      message: "Your mood has been trending downward. Consider reaching out to campus counseling services.",
      priority: 'high',
      actions: ['Schedule counseling appointment', 'Connect with peer support', 'Practice daily mindfulness']
    };
  } else if (trend === 'improving') {
    return {
      message: "Great progress! Keep up the positive momentum with your current wellness practices.",
      priority: 'low',
      actions: ['Maintain current routine', 'Share success with support network', 'Set new wellness goals']
    };
  } else {
    return {
      message: "Your mood has been stable. Consider adding new wellness activities to boost your mental health.",
      priority: 'medium',
      actions: ['Try new stress management techniques', 'Join a support group', 'Increase physical activity']
    };
  }
};

File: backend/app/services/ai_service.py
import asyncio
from typing import Dict, List, Any
import logging
import random
from datetime import datetime

logger = logging.getLogger(__name__)

class AIService:
    def __init__(self):
        self.models_loaded = False
        self.emotion_keywords = {
            'positive': ['happy', 'excited', 'grateful', 'amazing', 'wonderful', 'love', 'blessed', 'fantastic', 'awesome', 'great'],
            'negative': ['sad', 'depressed', 'anxious', 'worried', 'stressed', 'hate', 'terrible', 'awful', 'horrible', 'devastated'],
            'stress': ['overwhelmed', 'pressure', 'deadline', 'exam', 'finals', 'study', 'homework', 'project', 'assignment', 'busy']
        }

    async def initialize(self):
        """Initialize AI models (simulated for demo)"""
        logger.info("Initializing AI models...")
        # Simulate model loading time
        await asyncio.sleep(2)
        self.models_loaded = True
        logger.info("AI models loaded successfully")

    def is_ready(self) -> bool:
        return self.models_loaded

    async def analyze_sentiment(self, text: str) -> Dict[str, Any]:
        """Analyze sentiment of text"""
        if not self.models_loaded:
            raise Exception("AI models not loaded")

        # Simple keyword-based analysis for demo
        words = text.lower().split()
        
        positive_score = sum(1 for word in words if word in self.emotion_keywords['positive'])
        negative_score = sum(1 for word in words if word in self.emotion_keywords['negative'])
        stress_score = sum(1 for word in words if word in self.emotion_keywords['stress'])
        
        # Determine sentiment
        if stress_score > 0 and stress_score >= positive_score:
            label = 'stressed'
            confidence = min(0.7 + stress_score * 0.1, 0.95)
        elif negative_score > positive_score:
            label = 'negative'
            confidence = min(0.6 + negative_score * 0.1, 0.95)
        elif positive_score > negative_score:
            label = 'positive'
            confidence = min(0.6 + positive_score * 0.1, 0.95)
        else:
            label = 'neutral'
            confidence = 0.5

        return {
            'label': label,
            'confidence': confidence,
            'scores': {
                'positive': positive_score,
                'negative': negative_score,
                'stress': stress_score
            }
        }

    async def analyze_typing_patterns(self, keystroke_data: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Analyze typing patterns for emotional indicators"""
        if not keystroke_data:
            return {'emotional_state': 'neutral', 'confidence': 0.5}

        avg_speed = sum(k.get('speed', 0) for k in keystroke_data) / len(keystroke_data)
        long_pauses = len([k for k in keystroke_data if k.get('pause', 0) > 2000])
        rapid_typing = len([k for k in keystroke_data if k.get('speed', 0) > 5])

        emotional_state = 'neutral'
        confidence = 0.5

        if long_pauses > 3:
            emotional_state = 'contemplative'
            confidence = 0.7
        elif rapid_typing > 5:
            emotional_state = 'agitated'
            confidence = 0.8
        elif avg_speed < 2:
            emotional_state = 'tired'
            confidence = 0.6

        return {
            'emotional_state': emotional_state,
            'confidence': confidence,
            'metrics': {
                'average_speed': avg_speed,
                'long_pauses': long_pauses,
                'rapid_typing': rapid_typing
            }
        }

    async def analyze_social_media_content(self, url: str, content: str, platform: str, interaction_time: float) -> Dict[str, Any]:
        """Analyze social media content for emotional impact"""
        sentiment = await self.analyze_sentiment(content)
        
        # Simulate content impact analysis
        impact_score = random.uniform(0.3, 0.9)
        if sentiment['label'] == 'negative':
            impact_score += 0.2
        
        triggers = []
        if 'exam' in content.lower() or 'test' in content.lower():
            triggers.append('academic_stress')
        if 'lonely' in content.lower() or 'alone' in content.lower():
            triggers.append('social_isolation')

        return {
            'sentiment': sentiment,
            'impact_score': min(impact_score, 1.0),
            'triggers': triggers,
            'recommendation': self._get_content_recommendation(sentiment['label']),
            'platform': platform,
            'interaction_time': interaction_time
        }

    async def analyze_page_interaction(self, url: str, time_spent: float, scroll_pattern: List[float], click_pattern: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Analyze page interaction patterns"""
        distress_indicators = []
        
        # Rapid scrolling pattern
        if len(scroll_pattern) > 10 and time_spent < 300:  # Many scrolls in 5 minutes
            distress_indicators.append('rapid_scrolling')
        
        # Excessive clicking
        if len(click_pattern) > 20 and time_spent < 300:
            distress_indicators.append('excessive_clicking')
        
        # Long session time
        if time_spent > 7200:  # 2 hours
            distress_indicators.append('excessive_usage')

        return {
            'distress_indicators': distress_indicators,
            'engagement_score': min(time_spent / 1800, 1.0),  # Normalize to 30 minutes
            'interaction_intensity': len(click_pattern) / max(time_spent / 60, 1),  # Clicks per minute
            'recommendations': self._get_usage_recommendations(distress_indicators)
        }

    def _get_content_recommendation(self, sentiment: str) -> str:
        recommendations = {
            'negative': 'Consider taking a break from social media and try a breathing exercise.',
            'stressed': 'This content might be adding to your stress. Would you like to see some calming content instead?',
            'positive': 'Great to see positive content! Keep engaging with uplifting posts.',
            'neutral': 'Balance is key. Mix positive content with your current browsing.'
        }
        return recommendations.get(sentiment, 'Stay mindful of how content affects your mood.')

    def _get_usage_recommendations(self, indicators: List[str]) -> List[str]:
        recommendations = []
        if 'rapid_scrolling' in indicators:
            recommendations.append('Try slowing down your browsing pace')
        if 'excessive_clicking' in indicators:
            recommendations.append('Take breaks between interactions')
        if 'excessive_usage' in indicators:
            recommendations.append('Consider setting usage limits')
        return recommendations

    async def get_system_load(self) -> Dict[str, Any]:
        """Get current system load metrics"""
        return {
            'cpu_usage': random.uniform(20, 80),
            'memory_usage': random.uniform(30, 70),
            'active_analyses': random.randint(50, 200),
            'queue_length': random.randint(0, 10)
        }

    def get_model_status(self) -> Dict[str, str]:
        """Get status of AI models"""
        return {
            'sentiment_model': 'healthy' if self.models_loaded else 'loading',
            'typing_analysis': 'healthy' if self.models_loaded else 'loading',
            'pattern_recognition': 'healthy' if self.models_loaded else 'loading'
        }